tests/integration/relations/pgbouncer_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,199 @@
+        )
+        self.framework.observe(
+            self.database_clusters.on.database_created, self._on_cluster_database_created
+        )
+        self.framework.observe(
+            self.database_clusters.on.endpoints_changed,
+            self._on_cluster_endpoints_changed,
+        )
+
+        self.framework.observe(self.on.run_sql_action, self._on_run_sql_action)
+
+    def _on_start(self, _) -> None:
+        """Only sets an Active status."""
+        self.unit.status = ActiveStatus()
+
+    # First database events observers.
+    def _on_first_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"first database credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials of the first database")
+
+    def _on_first_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    # Second database events observers.
+    def _on_second_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"second database credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials of the second database")
+
+    def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    # Multiple database clusters events observers.
+    def _on_cluster_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(
+            f"cluster {event.relation.app.name} credentials: {event.username} {event.password}"
+        )
+        self.unit.status = ActiveStatus(
+            f"received database credentials for cluster {event.relation.app.name}"
+        )
+
+    def _on_cluster_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(
+            f"cluster {event.relation.app.name} endpoints have been changed to: {event.endpoints}"
+        )
+
+    # Multiple database clusters events observers (for aliased clusters/relations).
+    def _on_cluster1_database_created(self, event: DatabaseCreatedEvent) -> None:
```

Comment:
This and the other methods below are not needed, as they are not mentioned as callbacks for any events that are being observed.